benchmarks/lobsters/route_generator.rb

```diff
@@ -0,0 +1,121 @@
+# Generate a set of routes for Lobsters
+
+class RouteGenerator
+  # Take a variety of routes and randomise order, distribution and specific data items (comments, users.)
+  ROUTE_GROUPS = [
+    { num: 15, method: :GET, routes: ["/u"] }, # Users tree, showing order of invitation - lots of view logic
+    { num: 15, method: :GET, routes: ["/active", "/newest", "/recent", "/hottest"] }, # Views of the stories by attributes
+    { num: 8, method: :GET, routes: ["/rss", "/privacy", "/about", "/settings"] }, # Less-common and less-interesting routes for variation
+    { num: 8, method: :GET, routes: ["/top?length=1d", "/top?length=1w", "/top?length=1y"] }, # Top stories by time
+    { num: 15, method: :GET, routes: ["/hidden", "/saved", "/upvoted/stories"] }, # These all required being logged in
+
+    # Turned off flag_warning check for /threads -- to hard to port to SQLite
+    { num: 15, routes: ["/comments", "/upvoted/comments", "/threads", "/comments/:comment_id/reply"] },
+    { num: 8, routes: ["/threads/:username", "/u/:username"] },
+
+    { num: 15, routes: ["/replies", "/replies/comments", "/replies/stories", "/replies/unread"] },  # replies#stories, replies#unread, replies#comments
+
+    #{ num: 25, routes: ["/s/:story_id"] },  # need to check more into how stories work - e.g. story#show doesn't seem side-effect-free
```

Comment:
It seems a bit unfortunate that the /s/ route is commented out because I think this is the main thing that lobsters does right, show stories?

Can you investigate what are the side-effects?

Reply:
Most of the side effects aren't bad -- e.g. number-of-views tracking on stories. That's fine.

Somewhat harder is the complex "can this user see this story?" logic, and I assume we'd prefer not to just allow 404s.

We get a 404 if a story is deleted, or is downvoted sufficiently but not submitted by this user. But we can just use the existing can_be_seen_by_user? check in the benchmark and I think that'll clear all that up.

We could also switch to doing the benchmark with a moderator, who can see and do a lot more. But I'm worried that the changes to performance will be significant -- a *lot* of filtering logic changes for moderators.

So I think we can do stories#show that way. I don't want to do most other story operations (e.g. upvote/downvote or delete) since that can change the story visibility mid-benchmark, which makes the route randomisation problem a lot harder. We can't use the model can_be_seen_by_user? method to check DB state unless we mutate DB state during route randomisation.

If we decide to do a more complex version of this benchmark in future, that's the kind of thing we could add. In fact the in-memory database makes it easier -- load the DB into memory, go through and do the destructive operations while randomising routes and tracking N users, output the final route list, drop the in-mem DB and reload. We could even do the route randomisation as a pre-pass, like the DB faking. We'd still need route randomiser code to generate the initial routes, and for later changes.

So: I'll add the /s/:story_id route, filtering for stories that can be seen by our single benchmark user for now. Unless I hear otherwise, I assume that the more complex version of all of this can wait for the not-yet-scheduled future.

Reply:
Okay, basic /s/:story_id route is in.